text/0000-cargo-workspace.md

```diff
@@ -0,0 +1,345 @@
+  root/
+    Cargo.toml
+    src/
+  crates/
+    crate1/
+      Cargo.toml
+      src/
+    crate2/
+      Cargo.toml
+      src/
+  ```
+
+  The workspace here can be configured by placing the following in the
+  manifests:
+
+  ```toml
+  # root/Cargo.toml
+  #
+  # Note that `members` aren't necessary if these are otherwise path
+  # dependencies.
+  [workspace]
+  members = ["../crates/crate1", "../crates/crate2"]
+  ```
+
+  ```toml
+  # crates/crate1/Cargo.toml
+  [package]
+  workspace = "../root"
+  ```
+
+  ```toml
+  # crates/crate2/Cargo.toml
+  [package]
+  workspace = "../root"
+  ```
+
+Projects like the compiler will likely need exhaustively explicit configuration.
+The `rust` repo conceptually has two workspaces, the standard library and the
+compiler, and these would need to be manually configured with
+`workspace.members` and `package.workspace` keys amongst all crates.
+
+### Lockfile and override interactions
+
+One of the main features of a workspace is that only one `Cargo.lock` is
+generated for the entire workspace. This lock file can be affected, however,
+with both [`[replace]` overrides][replace] as well as `paths` overrides.
+
+[replace]: https://github.com/rust-lang/cargo/pull/2385
+
+Primarily, the `Cargo.lock` generate will not simply be the concatenation of the
+lock files from each project. Instead the entire workspace will be resolved
+together all at once, minimizing versions of crates used and sharing
+dependencies as much as possible. For example one `path` dependency will always
+have the same set of dependencies no matter which crate is being compiled.
+
+When interacting with overrides, workspaces will be modified to only allow
+`[replace]` to exist in the workspace root. This Cargo.toml will affect lock
+file generation, but no other workspace members will be allowed to have a
+`[replace]` directive (with an informative error message being produced).
+
+Finally, the `paths` overrides will be applied as usual, and they'll continue to
+be applied relative to whatever crate is being compiled (not the workspace
+root). These are intended for much more local testing, so no restriction of
+"must be in the root" should be necessary.
+
+Note that this change to the lockfile format is technically incompatible with
+older versions of Cargo.lock, but the entire workspaces feature is also
+incompatible with older versions of Cargo. This will require projects that wish
+to work with workspaces and multiple versions of Cargo to check in multiple
+`Cargo.lock` files, but if projects avoid workspaces then Cargo will remain
+forwards and backwards compatible.
+
+### Future Extensions
+
+Once Cargo understands a workspace of crates, we could easily extend various
+subcommands with a `--all` flag to perform tasks such as:
+
+* Test all crates within a workspace (run all unit tests, doc tests, etc)
+* Build all binaries for a set of crates within a workspace
+* Publish all crates in a workspace if necessary to crates.io
```

Comment:
Related to this use case, another possible extension is to share a version across all the crates in a workspace. Some projects act as a collection of crates that are published at the same time (e.g., `foo` and `foo-sys`) with the same version. It'd be a nice convenience to not have to update the `version` property in every `Cargo.toml`.